System.InvalidCastException for object of type 'Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.BindingAnalyzer'

Getting the following exception from the binlog of a project referencing the Microsoft.Azure.WebJobs.Extensions.DurableTask package:

```
[52184] [001] [INFO]  Running binary log analysis.
[52184] [001] [INFO] Reading compiler calls from binary log K:\dbs\sh\1f11\0210_192431_0\sources\dev\IPClassification\WorkerExtensions\Logs\Debug\Amd64\msbuild.binlog
[52184] [001] [INFO]   Found 1 compilations in binary log
[52184] [001] [ERROR]   Unhandled exception: System.InvalidCastException: Unable to cast object of type 'Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.BindingAnalyzer' to type 'Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer'.
   at Basic.CompilerLog.Util.Impl.<BasicAnalyzerHostInMemory>F9E8961216939B16257C8367E4503A5C59B8ADFCFB9230E2D4D147DC2CB1F9FA5__BasicAnalyzerReference.GetAnalyzers(Builder builder, String languageName) in /home/runner/work/complog/complog/src/Basic.CompilerLog.Util/Impl/BasicAnalyzerHostInMemory.cs:line 278
   at Basic.CompilerLog.Util.Impl.<BasicAnalyzerHostInMemory>F9E8961216939B16257C8367E4503A5C59B8ADFCFB9230E2D4D147DC2CB1F9FA5__BasicAnalyzerReference.GetAnalyzers(String language) in /home/runner/work/complog/complog/src/Basic.CompilerLog.Util/Impl/BasicAnalyzerHostInMemory.cs:line 227
   at Basic.CompilerLog.Util.CompilationData.EnsureAnalyzersLoaded() in /home/runner/work/complog/complog/src/Basic.CompilerLog.Util/CompilationData.cs:line 176
   at Basic.CompilerLog.Util.CompilationData.GetGenerators() in /home/runner/work/complog/complog/src/Basic.CompilerLog.Util/CompilationData.cs:line 115
   at Basic.CompilerLog.Util.CSharpCompilationData.CreateGeneratorDriver() in /home/runner/work/complog/complog/src/Basic.CompilerLog.Util/CompilationData.cs:line 374
   at Basic.CompilerLog.Util.CompilationData.GetCompilationAfterGenerators(ImmutableArray`1& diagnostics, CancellationToken cancellationToken) in /home/runner/work/complog/complog/src/Basic.CompilerLog.Util/CompilationData.cs:line 132
   at Basic.CompilerLog.Util.CompilationData.GetCompilationAfterGenerators(CancellationToken cancellationToken) in /home/runner/work/complog/complog/src/Basic.CompilerLog.Util/CompilationData.cs:line 120
   at Semmle.Extraction.CSharp.Extractor.RunBinaryLogAnalysis(Stopwatch stopwatch, Options options, String binlogPath, ILogger logger, CanonicalPathCache canonicalPathCache, PathTransformer pathTransformer)
   at Semmle.Extraction.CSharp.Extractor.RunBinaryLogAnalysis(Stopwatch stopwatch, Options options, String[] binlogPaths, ILogger logger, CanonicalPathCache canonicalPathCache, PathTransformer pathTransformer)
   at Semmle.Extraction.CSharp.Extractor.Run(String[] args)
```

I'm happy to share the binlog privately if needed.

--- a/.vscode/launch.json
+++ b/.vscode/launch.json
@@ -13,11 +13,9 @@
             "program": "${workspaceFolder}/artifacts/bin/Basic.CompilerLog/debug_net9.0/Basic.CompilerLog.dll",
             "args": [
                 "replay",
-                "/home/jaredpar/code/msbuild/artifacts/log/Debug/Build.binlog",
-                "-p",
-                "StringTools.UnitTests.csproj",
-                "-n"],
-            "cwd": "/home/jaredpar/code/msbuild/artifacts/log/Debug",
+                "C:/Users/jaredpar/temp/console/build.complog"
+            ],
+            "cwd": "${workspaceFolder}/artifacts/",
             // "cwd": "${workspaceFolder}/src/Basic.CompilerLog",
             // For more information about the 'console' field, see https://aka.ms/VSCode-CS-LaunchJson-Console
             "console": "internalConsole",

--- a/src/Basic.CompilerLog.UnitTests/Basic.CompilerLog.UnitTests.csproj
+++ b/src/Basic.CompilerLog.UnitTests/Basic.CompilerLog.UnitTests.csproj
@@ -20,7 +20,7 @@
     <PackageReference Include="xunit.v3" />
     <ProjectReference Include="..\Basic.CompilerLog.Util\Basic.CompilerLog.Util.csproj" />
     <ProjectReference Include="..\Basic.CompilerLog\Basic.CompilerLog.csproj" Condition="'$(TargetFramework)' == 'net8.0'" />
-    <PackageReference Include="Basic.Reference.Assemblies.Net60" />
+    <PackageReference Include="Basic.Reference.Assemblies.Net80" />
     <PackageReference Include="coverlet.msbuild">
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
       <PrivateAssets>all</PrivateAssets>

--- a/src/Basic.CompilerLog.UnitTests/BasicAnalyzerHostTests.cs
+++ b/src/Basic.CompilerLog.UnitTests/BasicAnalyzerHostTests.cs
@@ -81,7 +81,7 @@ public sealed class BasicAnalyzerHostTests : TestBase
         var compilation = CSharpCompilation.Create(
             "example",
             [],
-            Basic.Reference.Assemblies.Net60.References.All);
+            Basic.Reference.Assemblies.Net80.References.All);
         var driver = CSharpGeneratorDriver.Create([host.Generator!]);
         driver.RunGeneratorsAndUpdateCompilation(compilation, out var compilation2, out var diagnostics, CancellationToken);
         Assert.Empty(diagnostics);

--- a/src/Basic.CompilerLog.UnitTests/InMemoryLoaderTests.cs
+++ b/src/Basic.CompilerLog.UnitTests/InMemoryLoaderTests.cs
@@ -1,8 +1,13 @@
 namespace Basic.CompilerLog.UnitTests;
 
 using System.Runtime.InteropServices;
+#if NET
+using System.Runtime.Loader;
+#endif
 using Basic.CompilerLog.Util;
+using Basic.CompilerLog.Util.Impl;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
 using Xunit;
 
 [Collection(CompilerLogCollection.Name)]
@@ -51,5 +56,25 @@ public sealed class InMemoryLoaderTests : TestBase
         var specific = generators.Where(x => TestUtil.GetGeneratorType(x).Name == analyzerTypeName);
         Assert.Equal(expectedCount, specific.Count());
     }
+
+    [Fact]
+    public void AnalyzersBadDefinition()
+    {
+        var (fileName, image) = LibraryUtil.GetAnalyzersWithBadMetadata();
+        var alc = new AssemblyLoadContext("Custom", isCollectible: true);
+        var loader = new InMemoryLoader(
+            "test loader",
+            alc,
+            Path.GetFileNameWithoutExtension(fileName),
+            image.ToArray(),
+            d => { Assert.Fail(d.GetMessage()); });
+        var analyzerReference = loader.AnalyzerReferences.Single();
+        var analyzer = analyzerReference.GetAnalyzersForAllLanguages().Single();
+        Assert.Equal("GoodAnalyzer", analyzer.GetType().Name);
+        var generator = analyzerReference.GetGeneratorsForAllLanguages().Single();
+        Assert.Equal("GoodGenerator", TestUtil.GetGeneratorType(generator).Name);
+        loader.Unload();
+    }
+
 #endif
 }

--- a/src/Basic.CompilerLog.UnitTests/LibraryUtil.cs
+++ b/src/Basic.CompilerLog.UnitTests/LibraryUtil.cs
@@ -1,4 +1,5 @@
 
+using System.Drawing.Text;
 using System.Text;
 using Basic.Reference.Assemblies;
 using Microsoft.CodeAnalysis;
@@ -40,10 +41,57 @@ internal static class LibraryUtil
 
         var compilation = CSharpCompilation.Create(
             "SimplePia",
-            [ CSharpSyntaxTree.ParseText(content1), CSharpSyntaxTree.ParseText(content2) ],
-            Net60.References.All,
+            [CSharpSyntaxTree.ParseText(content1), CSharpSyntaxTree.ParseText(content2)],
+            Net80.References.All,
             new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
 
+        return Compile(compilation, "SimplePia.dll");
+    }
+
+    /// <summary>
+    /// Returns an assembly that has a set of well defined analyzers and generators and ones
+    /// with definition errors.
+    /// </summary>
+    internal static (string FileName, MemoryStream Image) GetAnalyzersWithBadMetadata()
+    {
+        var code = """
+            using System.Collections.Immutable;
+            using Microsoft.CodeAnalysis;
+            using Microsoft.CodeAnalysis.Diagnostics;
+
+            [DiagnosticAnalyzer(LanguageNames.CSharp)]
+            public class BadAnalyzer
+            {
+            }
+
+            [DiagnosticAnalyzer(LanguageNames.CSharp)]
+            public class GoodAnalyzer : DiagnosticAnalyzer
+            {
+                public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray<DiagnosticDescriptor>.Empty;
+                public override void Initialize(AnalysisContext context) { }
+            }
+
+            [Generator(LanguageNames.CSharp)]
+            public class BadGenerator { }
+
+            [Generator(LanguageNames.CSharp)]
+            public class GoodGenerator : IIncrementalGenerator
+            {
+                public void Initialize(IncrementalGeneratorInitializationContext context) { }
+            }
+            """;
+
+        var roslynReference = MetadataReference.CreateFromFile(typeof(Compilation).Assembly.Location);
+        var compilation = CSharpCompilation.Create(
+            "TestAnalyzers",
+            [CSharpSyntaxTree.ParseText(code)],
+            [.. Net80.References.All, roslynReference],
+            new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+        return Compile(compilation, "TestAnalyzers.dll");
+    }
+
+    private static (string FileName, MemoryStream Image) Compile(Compilation compilation, string fileName)
+    {
         var peStream = new MemoryStream();
         var emitResult = compilation.Emit(peStream);
         if (!emitResult.Success)
@@ -52,7 +100,7 @@ internal static class LibraryUtil
         }
 
         peStream.Position = 0;
-        return ("SimplePia.dll", peStream);
+        return (fileName, peStream);
 
         string GetMessage(IEnumerable<Diagnostic> diagnostics)
         {
@@ -64,7 +112,5 @@ internal static class LibraryUtil
             }
             return builder.ToString();
         }
-
     }
-
 }

--- a/src/Basic.CompilerLog.Util/Impl/BasicAnalyzerHostInMemory.cs
+++ b/src/Basic.CompilerLog.Util/Impl/BasicAnalyzerHostInMemory.cs
@@ -58,6 +58,15 @@ internal sealed class InMemoryLoader : AssemblyLoadContext
         AnalyzerReferences = builder.MoveToImmutable();
     }
 
+    internal InMemoryLoader(string name, AssemblyLoadContext compilerLoadContext, string simpleName, byte[] bytes, Action<Diagnostic> onDiagnostic)
+        :base(name, isCollectible: true)
+    {
+        CompilerLoadContext = compilerLoadContext;
+        _map[simpleName] = bytes;
+        var reference = new BasicAnalyzerReference(new AssemblyName(simpleName), bytes, this, onDiagnostic);
+        AnalyzerReferences = [reference];
+    }
+
     protected override Assembly? Load(AssemblyName assemblyName)
     {
         try
@@ -288,7 +297,10 @@ file sealed class BasicAnalyzerReference : AnalyzerReference
         var attributeType = typeof(DiagnosticAnalyzerAttribute);
         foreach (var type in GetTypes(attributeType.Namespace!, attributeType.Name, languageName, CoreAction))
         {
-            builder.Add((DiagnosticAnalyzer)CreateInstance(type));
+            if (Activator.CreateInstance(type) is DiagnosticAnalyzer d)
+            {
+                builder.Add(d);
+            }
         }
 
         static void CoreAction(
@@ -326,14 +338,13 @@ file sealed class BasicAnalyzerReference : AnalyzerReference
         var attributeType = typeof(GeneratorAttribute);
         foreach (var type in GetTypes(attributeType.Namespace!, attributeType.Name, languageName, CoreAction))
         {
-            var generator = CreateInstance(type);
+            var generator = Activator.CreateInstance(type);
             if (generator is ISourceGenerator sg)
             {
                 builder.Add(sg);
             }
-            else
+            else if (generator is IIncrementalGenerator ig)
             {
-                IIncrementalGenerator ig = (IIncrementalGenerator)generator;
                 builder.Add(ig.AsSourceGenerator());
             }
         }
@@ -371,14 +382,6 @@ file sealed class BasicAnalyzerReference : AnalyzerReference
                 }
             }
         }
-    }
-
-    private static object CreateInstance(Type type)
-    {
-        var ctor = type.GetConstructors()
-            .Where(c => c.GetParameters().Length == 0)
-            .Single();
-        return ctor.Invoke(null);
     }
 
     public override string ToString() => $"In Memory {AssemblyName}";

--- a/src/Directory.Packages.props
+++ b/src/Directory.Packages.props
@@ -5,7 +5,7 @@
     <_RoslynVersion>4.12.0</_RoslynVersion>
   </PropertyGroup>
   <ItemGroup>
-    <PackageVersion Include="Basic.Reference.Assemblies.Net60" Version="1.8.0" />
+    <PackageVersion Include="Basic.Reference.Assemblies.Net80" Version="1.8.0" />
     <PackageVersion Include="BenchmarkDotNet" Version="0.14.0" />
     <PackageVersion Include="BenchmarkDotNet.Diagnostics.Windows" Version="0.13.5" />
     <PackageVersion Include="coverlet.msbuild" Version="6.0.4" />

--- a/src/Scratch/Scratch.cs
+++ b/src/Scratch/Scratch.cs
@@ -386,7 +386,7 @@ static void RoslynScratch()
     var compilation = CSharpCompilation.Create(
         "scratch",
         new[] { syntaxTree },
-        Basic.Reference.Assemblies.Net60.References.All);
+        Basic.Reference.Assemblies.Net80.References.All);
 
     var context = compilation.GetSemanticModel(syntaxTree);
     var token = syntaxTree

--- a/src/Scratch/Scratch.csproj
+++ b/src/Scratch/Scratch.csproj
@@ -18,7 +18,7 @@
     <ProjectReference Include="..\Basic.CompilerLog\Basic.CompilerLog.csproj" />
     <PackageReference Include="BenchmarkDotNet" />
     <PackageReference Include="BenchmarkDotNet.Diagnostics.Windows" />
-    <PackageReference Include="Basic.Reference.Assemblies.Net60" />
+    <PackageReference Include="Basic.Reference.Assemblies.Net80" />
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Features" />
     <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic.Features" />
   </ItemGroup>

